More mapgen sanity checking and control

jbytheway opened the pull request:
#### Summary
Infrastructure "Better checking of mapgen definitions"

#### Purpose of change
We have a lot of mapgen that runs on top of existing maps.  This can happen due to nested chunks, map extras, predecessor mapgen, update mapgen, or some hardcoded things.

#### Describe the solution
This PR adds a sanity check for whenever mapgen changes the terrain on a tile where something (furniture, trap, or item) already exists.  This is often a sign of a problem.  When it happens, it flags a warning message.

It also provides a variety of solutions to this problem that mappers can use:
- Two new mapgen flags `ALLOW_TERRAIN_UNDER_OTHER_DATA` and `ERASE_ALL_BEFORE_PLACING_TERRAIN` which specify the desired default behaviour if a particular piece of mapgen wishes to use the same behaviour whenever this happens.  The first flag suppresses the warning while the second causes everything on the tile to be removed before setting the terrain (which is normally what you want).
- A new `remove_all` mapgen command that erases everything on a tile, for when the mapper needs to be more targeted in removing only in some cases.

The recently-added exhaustive mapgen unit test allows these problems to be quickly discovered, so I ran that a lot and fixed all the issues it found (although, as usual, there's probably a long tail of low-probability issues I didn't fix).

I only ran the tests for Magiclysm, not other mods, so there are probably issues there too.

I still need to write documentation, so opening this as a draft.

#### Describe alternatives you've considered
I considered enhancing the `terrain` definition itself to allow specifying how to handle this situation.  That might be better than adding the `remove_all` command for fixing these issues, but I think `remove_all` is probably something we want to have regardless.

#### Testing
Running the unit tests a lot.

#### Additional context
I started this work back in October, but realised it was infeasible to fix all the issues without having a unit test to help, so I have been diverted by adding that unit test in the time since and only recently returned to this.

Night-Pryanik commented on src/mapgen.cpp:
```diff
@@ -2942,6 +2993,33 @@ class jmapgen_remove_vehicles : public jmapgen_piece
+                if( optional_vpart_position vp = dat.m.veh_at( p ) ) {
+                    dat.m.destroy_vehicle( &vp->vehicle() );
+                    get_map().clear_vehicle_level_caches();
+                }
```
Graffiti could be removed as well.

jbytheway replied:
Done

PatrikLundell commented:
This is probably better for "real" map gen than setting the palettes to clear furniture, and it adds the removal of items.

You probably don't want to remove all items for camp constructions, though, as that risks screwing the player over when the new construction is built on top of collected loot (and the player doesn't get any indication where the construction will be placed, although the version 2 sites tries to provide a limited indication). Also, you don't want a 100 page spam report of every item in every tile that's happen to be in the construction location.

Thus, this seems to be a useful addition that should be used with a little bit of care.

jbytheway commented:
> This is probably better for "real" map gen than setting the palettes to clear furniture, and it adds the removal of items.

Indeed.  In some cases I've removed the explicit clearing of furniture where it is replaced by this new feature.

> You probably don't want to remove all items for camp constructions, though, as that risks screwing the player over when the new construction is built on top of collected loot (and the player doesn't get any indication where the construction will be placed, although the version 2 sites tries to provide a limited indication).

I haven't touched camp constructions.  They need a different approach to fix them.

> Also, you don't want a 100 page spam report of every item in every tile that's happen to be in the construction location.

They probably will lead to some spam in the current state of things, but it would only be one report per tile, not one per item.

> Thus, this seems to be a useful addition that should be used with a little bit of care.

Thanks :).

jbytheway commented:
Added docs and fixed a few more issues.  Tests fail rarely now (I had about a hundred successes in a row at best).

Marking this ready for review now, but I should note that I haven't fixed any issues in mods except for Magiclysm.  Other mods (in-repo and otherwise) will probably trigger this error quite often.  But it's only a `debugmsg`; if players hit `i` to ignore it then it won't bother them again for that session.

kevingranade commented:
Just rebased (no conflicts) to poke the build.

jbytheway commented:
Looks like it finally got through most of the tests and the one remaining failure is unrelated.
